src/Compiler/Utilities/Cancellable.fs

```diff
@@ -0,0 +1,226 @@
+namespace FSharp.Compiler
+
+open System
+open System.Threading
+open Internal.Utilities.Library
+
+[<Sealed>]
+type Cancellable =
+    [<ThreadStatic; DefaultValue>]
+    static val mutable private tokens: CancellationToken list
+
+    static let disposable =
+        { new IDisposable with
+            member this.Dispose() =
+                Cancellable.Tokens <- Cancellable.Tokens |> List.tail
+        }
+
+    static member Tokens
+        with private get () =
+            match box Cancellable.tokens with
+            | Null -> []
+            | _ -> Cancellable.tokens
+        and private set v = Cancellable.tokens <- v
+
+    static member UsingToken(ct) =
+        Cancellable.Tokens <- ct :: Cancellable.Tokens
+        disposable
+
+    static member Token =
+        match Cancellable.Tokens with
+        | [] -> CancellationToken.None
+        | token :: _ -> token
+
+    static member CheckAndThrow() =
+        match Cancellable.Tokens with
+        | [] -> ()
+        | token :: _ -> token.ThrowIfCancellationRequested()
+
+[<AutoOpen>]
+module Cancellable =
+    type Exception with
+
+        member this.IsOperationCancelled =
+            match this with
+            | :? OperationCanceledException -> true
+            | _ -> false
+
+namespace Internal.Utilities.Library
+
+open System
+open System.Threading
+open FSharp.Compiler
+
+#if !FSHARPCORE_USE_PACKAGE
+open FSharp.Core.CompilerServices.StateMachineHelpers
+#endif
+
+[<RequireQualifiedAccess; Struct>]
+type ValueOrCancelled<'TResult> =
```

Comment:
Do we want it to be comparable? Probably only need equality, right?

Reply:
Maybe not even equality, on the second thought.

Reply:
I've only moved this piece from another place. 🙂